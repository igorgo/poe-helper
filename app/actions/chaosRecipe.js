// @flow
import type { ChaosRecipeAction, ChaosRecipeItemType } from '../reducers/types';

export const INC_CHAOS_RECIPE = 'INC_CHAOS_RECIPE';
export const DEC_CHAOS_RECIPE = 'DEC_CHAOS_RECIPE';
export const RESET_CHAOS_RECIPE = 'RESET_CHAOS_RECIPE';
export const MINUS_SET_CHAOS_RECIPE = 'MINUS_SET_CHAOS_RECIPE';

export function increment(itemType: ChaosRecipeItemType): ChaosRecipeAction {
  return {
    type: INC_CHAOS_RECIPE,
    payload: {
      itemType
    }
  };
}

export function decrement(itemType: ChaosRecipeItemType): ChaosRecipeAction {
  return {
    type: DEC_CHAOS_RECIPE,
    payload: {
      itemType
    }
  };
}

export function reset(): ChaosRecipeAction {
  return {
    type: RESET_CHAOS_RECIPE
  };
}

export function minusSet(): ChaosRecipeAction {
  return {
    type: MINUS_SET_CHAOS_RECIPE
  };
}
