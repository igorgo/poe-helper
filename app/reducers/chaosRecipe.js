// @flow
import { remote } from 'electron';
import {
  INC_CHAOS_RECIPE,
  DEC_CHAOS_RECIPE,
  RESET_CHAOS_RECIPE,
  MINUS_SET_CHAOS_RECIPE
} from '../actions/chaosRecipe';
import type { ChaosRecipeAction, ChaosRecipeStore } from './types';
import { rows } from '../components/ChaosRecipe/ChaosRecipe';

const defaultState: ChaosRecipeStore = {
  weapon: 0,
  helmet: 0,
  body: 0,
  glove: 0,
  boot: 0,
  ring: 0,
  amulet: 0,
  belt: 0
};

const initialState: ChaosRecipeStore =
  (remote && remote.getGlobal('pstore').get('ChaosRecipe')) || defaultState;

const minusSet = (state: ChaosRecipeStore): ChaosRecipeStore => {
  const newState = {};
  Object.keys(state).forEach(i => {
    const newCount = state[i] - rows[i].length / 6;
    newState[i] = newCount < 0 ? 0 : newCount;
  });
  return newState;
};

export default function chaosRecipe(
  state: ChaosRecipeStore = initialState,
  action: ChaosRecipeAction
) {
  const itemType = action.payload ? action.payload.itemType : undefined;
  let newState = state;
  switch (action.type) {
    case INC_CHAOS_RECIPE:
      if (itemType && state[itemType] !== rows[itemType].length) {
        newState = { ...state, [itemType]: state[itemType] + 1 };
      }
      break;
    case DEC_CHAOS_RECIPE:
      if (itemType && state[itemType] !== 0) {
        newState = { ...state, [itemType]: state[itemType] - 1 };
      }
      break;
    case RESET_CHAOS_RECIPE:
      newState = defaultState;
      break;
    case MINUS_SET_CHAOS_RECIPE:
      newState = minusSet(state);
      break;
    default:
      newState = state;
  }
  if (remote) {
    remote.getGlobal('pstore').set('ChaosRecipe', newState);
  }
  return newState;
}
