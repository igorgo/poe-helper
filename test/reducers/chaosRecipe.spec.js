import chaosRecipe from '../../app/reducers/chaosRecipe';
import {
  RESET_CHAOS_RECIPE,
  DEC_CHAOS_RECIPE,
  INC_CHAOS_RECIPE,
  MINUS_SET_CHAOS_RECIPE
} from '../../app/actions/chaosRecipe';

const state = {
  weapon: 0,
  helmet: 0,
  body: 1,
  glove: 0,
  boot: 0,
  ring: 0,
  amulet: 0,
  belt: 0
};

describe('reducers', () => {
  describe('chaosRecipe', () => {
    it('should handle initial state', () => {
      expect(chaosRecipe(undefined, {})).toMatchSnapshot();
    });

    it('should handle INC_CHAOS_RECIPE', () => {
      expect(
        chaosRecipe(state, {
          type: INC_CHAOS_RECIPE,
          payload: { itemType: 'body' }
        })
      ).toMatchSnapshot();
    });

    it('should handle DEC_CHAOS_RECIPE', () => {
      expect(
        chaosRecipe(state, {
          type: DEC_CHAOS_RECIPE,
          payload: { itemType: 'body' }
        })
      ).toMatchSnapshot();
    });

    it('should handle MINUS_SET_CHAOS_RECIPE', () => {
      expect(
        chaosRecipe(state, { type: MINUS_SET_CHAOS_RECIPE })
      ).toMatchSnapshot();
    });

    it('should handle RESET_CHAOS_RECIPE', () => {
      expect(
        chaosRecipe(state, { type: RESET_CHAOS_RECIPE })
      ).toMatchSnapshot();
    });

    it('should handle unknown action type', () => {
      expect(chaosRecipe(state, { type: 'unknown' })).toMatchSnapshot();
    });
  });
});
