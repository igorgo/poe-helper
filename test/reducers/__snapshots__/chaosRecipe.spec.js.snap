// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers chaosRecipe should handle DEC_CHAOS_RECIPE 1`] = `
Object {
  "amulet": 0,
  "belt": 0,
  "body": 0,
  "boot": 0,
  "glove": 0,
  "helmet": 0,
  "ring": 0,
  "weapon": 0,
}
`;

exports[`reducers chaosRecipe should handle INC_CHAOS_RECIPE 1`] = `
Object {
  "amulet": 0,
  "belt": 0,
  "body": 2,
  "boot": 0,
  "glove": 0,
  "helmet": 0,
  "ring": 0,
  "weapon": 0,
}
`;

exports[`reducers chaosRecipe should handle MINUS_SET_CHAOS_RECIPE 1`] = `
Object {
  "amulet": 0,
  "belt": 0,
  "body": 0,
  "boot": 0,
  "glove": 0,
  "helmet": 0,
  "ring": 0,
  "weapon": 0,
}
`;

exports[`reducers chaosRecipe should handle RESET_CHAOS_RECIPE 1`] = `
Object {
  "amulet": 0,
  "belt": 0,
  "body": 0,
  "boot": 0,
  "glove": 0,
  "helmet": 0,
  "ring": 0,
  "weapon": 0,
}
`;

exports[`reducers chaosRecipe should handle initial state 1`] = `
Object {
  "amulet": 0,
  "belt": 0,
  "body": 0,
  "boot": 0,
  "glove": 0,
  "helmet": 0,
  "ring": 0,
  "weapon": 0,
}
`;

exports[`reducers chaosRecipe should handle unknown action type 1`] = `
Object {
  "amulet": 0,
  "belt": 0,
  "body": 1,
  "boot": 0,
  "glove": 0,
  "helmet": 0,
  "ring": 0,
  "weapon": 0,
}
`;
